apollo-federation/src/sources/connect/validation/errors.rs

```diff
@@ -0,0 +1,370 @@
+        coordinate: ErrorsCoordinate<'schema>,
+        schema: &'schema SchemaInfo,
+    ) -> Result<Option<Self>, Message> {
+        let Some((_, value)) = errors_arg
+            .iter()
+            .find(|(name, _)| name == &ERRORS_EXTENSIONS_ARGUMENT_NAME)
+        else {
+            return Ok(None);
+        };
+        let coordinate = ErrorsExtensionsCoordinate { coordinate };
+
+        // Ensure that the message selection is a valid JSON selection string
+        let string = match GraphQLString::new(value, &schema.sources) {
+            Ok(selection_str) => selection_str,
+            Err(_) => {
+                return Err(Message {
+                    code: Code::GraphQLError,
+                    message: format!("{coordinate} must be a string."),
+                    locations: value
+                        .line_column_range(&schema.sources)
+                        .into_iter()
+                        .collect(),
+                });
+            }
+        };
+        let selection = match JSONSelection::parse(string.as_str()) {
+            Ok(selection) => selection,
+            Err(err) => {
+                return Err(Message {
+                    code: Code::InvalidErrorsExtensions,
+                    message: format!("{coordinate} is not valid: {err}"),
+                    locations: value
+                        .line_column_range(&schema.sources)
+                        .into_iter()
+                        .collect(),
+                });
+            }
+        };
+        if selection.is_empty() {
+            return Err(Message {
+                code: Code::InvalidErrorsExtensions,
+                message: format!("{coordinate} is empty"),
+                locations: value
+                    .line_column_range(&schema.sources)
+                    .into_iter()
+                    .collect(),
+            });
+        }
+
+        Ok(Some(Self {
+            selection,
+            string,
+            coordinate,
+        }))
+    }
+
+    /// Check that the selection of the body matches the inputs at this location.
+    ///
+    /// TODO: check keys here?
+    pub(super) fn type_check(self, schema: &SchemaInfo) -> Result<(), Message> {
+        let Self {
+            selection,
+            string,
+            coordinate,
+        } = self;
+        let context = match coordinate.coordinate {
+            ErrorsCoordinate::Source { .. } => {
+                &Context::for_source_response(schema, &string, Code::InvalidErrorsMessage)
+            }
+            ErrorsCoordinate::Connect { connect } => {
+                &Context::for_connect_response(schema, connect, &string, Code::InvalidErrorsMessage)
+            }
+        };
+
+        expression::validate(
+            &Expression {
+                expression: selection,
+                location: 0..string.as_str().len(),
+            },
+            context,
+            &Shape::empty_object([]),
```

Comment:
I'm actually surprised that this works... I guess an object with properties satisfies an empty object? But I would expect us to be able to type-check that something is _explicitly_ empty in the future if we want to. I think the values here can be _anything_, so maybe this would be more accurate?

```suggestion
            &Shape::dict(Shape::unknown([]), []),
```